securesystemslib/metadata.py

```diff
@@ -97,3 +101,85 @@ def pae(self) -> bytes:
             len(self.payload),
             self.payload,
         )
+
+    def sign(self, signer: Signer) -> Signature:
+        """Sign the payload and create the signature.
+
+        Arguments:
+            signer: A "Signer" class instance.
+
+        Raises:
+            TypeError: If "signer" is not an instance of the "Signer" class.
+
+        Returns:
+            A "Signature" instance.
+        """
+
+        if not isinstance(signer, Signer):
+            raise TypeError(f"expected type 'Signer', got {type(signer).__name__}")
+
+        signature = signer.sign(self.pae)
+        self.signatures.append(signature)
+
+        return signature
+
+    def verify(self, keys: KeyList, threshold: int) -> List[str]:
+        """Verify the payload with the provided Keys.
+
+        Arguments:
+            keys: A list of a "str" and "Key" class instance.
+            threshold: Number of signatures needed to pass the verification.
+
+        Raises:
+            TypeError: If a key in "keys" is not an instance of the "Key" class.
+            ValueError: If "threshold" is not valid.
+            SignatureVerificationError: If the amount of "recognized_signers" is less
+                than provided threshold.
+
+        Returns:
+            recognized_signers: list of key names for which verification succeeds.
+        """
+
+        recognized_signers = []
+        used_keyids = []
+        pae = self.pae
+
+        # checks for threshold value.
+        if len(keys) < threshold or threshold <= 0:
+            raise ValueError(f"Threshold must be between 0 and {len(keys)}")
+
+        for signature in self.signatures:
+            for (name, key) in keys:
+                if not isinstance(key, Key):
+                    raise TypeError(f"expected type 'Key', got {type(key).__name__}")
+
+                # If key and signature include keyIDs but do not match skip.
+                if (
+                    signature.keyid is not None
+                    and key.keyid is not None
+                    and signature.keyid != key.keyid
+                ):
+                    continue
+
+                # If a key verifies the signature, we exit and use the result.
+                if key.verify(signature, pae):
+                    if key.keyid:
+                        # If keyid has already been verified, skip.
+                        if key.keyid in used_keyids:
```

Comment:
Let's not use the `keyid` to de-duplicate keys. The dsse spec makes it explicit that keyids must not be used for security decisions.

The way I read the [spec](https://github.com/secure-systems-lab/dsse/blob/469c6e9fa6c4b7252fb71101084561cfc4cd0fa5/protocol.md#multi-signature-verification) ...

> A (t, n)-ENVELOPE is valid if the enclosed signatures pass the verification against at least t of n **unique trusted** public keys.

... we need to trust the caller to pass a list of "unique trusted" keys. So I would say you can remove the `used_keyids` list and just check if the name is already in `recognized_signers` (btw. using a `set` might make this a bit faster). What do you think?


Might be a good idea to check with the dsse spec authors, if this assumption is correct. Mind creating an issue in the [spec repo](https://github.com/secure-systems-lab/dsse/issues)?

Either way we should definitely mention this expectation regarding the passed key list in the docstring.

Reply:
I took it from [dsse golang implementation](https://github.com/secure-systems-lab/go-securesystemslib/blob/c1e41ba5f168bb0cecaac7ce1bbe787c20e72d9b/dsse/verify.go#L93-L97), this is related to gpg subkeys.

Reply:
Ok, I will remove this piece of code.

> Can you see why?
GPG implementation is not done yet, and thats why I am not sure about this.

Reply:
Okay, I'm trying to say that the question of trust, realized via keyids, is not in the scope of DSSE. This is something that the caller should deal with. 

If I understand the spec correctly, DSSE just receives a list of public keys for a set of signatures, then it iterates over the signatures and tries to verify each signature, with each key, unless the key has been used before. If a threshold of signatures has been verified, we return success.

DSSE itself does not need to check if the keys are really distinct or if they are really trusted. It also does not need to check if the signatures are really distinct. This should happen implicitly when using distinct keys and using each key only once.